File: content/snippets/typescript/interview_lcci/03/01_three_in_one.ts
class TripleInOne {
  private readonly size: number;
  private readonly data: number[];
  private readonly tops: number[];

  constructor(stackSize: number) {
    this.size = stackSize;
    this.data = new Array(3 * stackSize).fill(0);
    this.tops = [0, 0, 0];
  }

  push(stackNum: number, value: number): void {
    if (this.tops[stackNum] === this.size) {
      return;
    }
    const index = stackNum * this.size + this.tops[stackNum];
    this.data[index] = value;
    this.tops[stackNum] += 1;
  }

  pop(stackNum: number): number {
    if (this.tops[stackNum] === 0) {
      return -1;
    }
    this.tops[stackNum] -= 1;
    const index = stackNum * this.size + this.tops[stackNum];
    return this.data[index];
  }

  peek(stackNum: number): number {
    if (this.tops[stackNum] === 0) {
      return -1;
    }
    const index = stackNum * this.size + this.tops[stackNum] - 1;
    return this.data[index];
  }

  isEmpty(stackNum: number): boolean {
    return this.tops[stackNum] === 0;
  }
}

function main(): void {
  const stacks = new TripleInOne(2);
  stacks.push(0, 10);
  stacks.push(0, 11);
  stacks.push(0, 12);
  stacks.push(1, 20);
  console.log(stacks.peek(0));
  console.log(stacks.pop(0));
  console.log(stacks.pop(0));
  console.log(stacks.pop(0));
  console.log(stacks.isEmpty(1));
}

main();

File: content/snippets/cpp/interview_lcci/03/01_three_in_one.cpp
#include <iostream>
#include <vector>

class TripleInOne {
public:
    TripleInOne(int stackSize)
        : size(stackSize), data(3 * stackSize, 0), tops(3, 0) {}

    void push(int stackNum, int value) {
        if (isFull(stackNum)) return;
        int index = stackNum * size + tops[stackNum];
        data[index] = value;
        tops[stackNum]++;
    }

    int pop(int stackNum) {
        if (isEmpty(stackNum)) return -1;
        tops[stackNum]--;
        int index = stackNum * size + tops[stackNum];
        return data[index];
    }

    int peek(int stackNum) const {
        if (isEmpty(stackNum)) return -1;
        int index = stackNum * size + tops[stackNum] - 1;
        return data[index];
    }

    bool isEmpty(int stackNum) const {
        return tops[stackNum] == 0;
    }

private:
    int size;
    std::vector<int> data;
    std::vector<int> tops;

    bool isFull(int stackNum) const {
        return tops[stackNum] == size;
    }
};

int main() {
    TripleInOne stacks(2);
    stacks.push(0, 10);
    stacks.push(0, 11);
    stacks.push(0, 12);
    stacks.push(1, 20);
    std::cout << stacks.peek(0) << '\n';
    std::cout << stacks.pop(0) << '\n';
    std::cout << stacks.pop(0) << '\n';
    std::cout << stacks.pop(0) << '\n';
    std::cout << std::boolalpha << stacks.isEmpty(1) << '\n';
    return 0;
}
